Interprocedural memref layout normalization

avarmapml opened the pull request:
Created an interprocedural pass to normalize memref layout.
In this pass all interprocedural memrefs with non-trivial layout maps are converted to ones with trivial identity layout ones.

avarmapml committed:
[MLIR] MemRef layout normalization at block 0

-- normalizeMemRef and SimplifyAffineStructure pass doesn't normalize memRefs
defined as function arguments.
-- This commit normalizes normalizes block 0's argument memRef layout
-- Function signature needs to match with the Block 0, this commit doesn't handle that

avarmapml committed:
[MLIR] Normalize non-dereferencing memref layout and modify function signature

-- This commit normalizes memref layouts at all the non-dereferencing memref uses
-- It modifies the function signature to match with Block 0's argument type and normalizes the
   memref layout at return type

avarmapml committed:
[MLIR] Interprocedural memref layout normalization

-- This commit creates a pass that converts all memref layout having non-trivial
   layout maps to one having trivial identity layout maps
-- Run using: --module-normalize-memref

avarmapml committed:
[MLIR] Interprocedural memref layout normalization

-- This commit creates a pass that converts all memref layout having non-trivial
   layout maps to one having trivial identity layout maps
-- Run using: --module-normalize-memref

avarmapml committed:
[MLIR] Interprocedural memref layout normalization

-- This commit creates a pass that converts all memref layout having non-trivial
   layout maps to one having trivial identity layout maps
-- Run using: --module-normalize-memref
-- Example input/output is shown in ModuleNormalizeMemRef.cpp

bondhugula commented on mlir/include/mlir/Dialect/Affine/Passes.h:
```diff
@@ -31,6 +31,10 @@ template <typename T> class OperationPass;
 /// layout map.
 std::unique_ptr<OperationPass<FuncOp>> createSimplifyAffineStructuresPass();
 
+// Creates a pass to normalize interprocedural memrefs to have the trivial 
```
Triple / comments here: `///`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+#include "llvm/Support/Format.h"
+#include "llvm/Support/raw_ostream.h"
+
+
```
Nit: drop extra new line.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+        // TODO(avarmapml): Check - A function's return type might have a different memref layout
+        // and a map.
+        for (auto retIndex : llvm::seq<unsigned>(0, funcOp.getNumResults())) {
+          if(results[retIndex] == (oldMemRef->getType())) {
```
Space after `if`.  Please use `clang-format`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+          .setShape(newShape)
+          .setAffineMaps(b.getMultiDimIdentityMap(newRank));
+
+        auto oldMemRef = new Value(funcOp.getArgument(argIndex));
```
You'll never need `new Value` in MLIR :-) `Value` is just a wrapper around the actual pointer, and should always be treated like a POD. Just use `Value oldMemRef = funcOp.getArgument(...`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+        inputs.push_back(static_cast<Type>(newMemRefType));
+        // break;
+      } else {
+        // Any other argument type can simply be part of the final function signature
```
Here and everywhere else. All comments should be terminated with a full stop and should be complete sentences.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -69,11 +69,12 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
     // escapes); no replacement in these cases.
-    return failure();
-
+    // return failure();
```
Remove commented code.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -92,119 +93,134 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
-      affineApplyOps.push_back(afOp);
+  OperationState state(op->getLoc(), op->getName());
+
+  // The following checks if op is dereferencing memref and performs the access index rewrites
```
Overflowing 80 columns. Please use `clang-format`.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -92,119 +93,134 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
-      affineApplyOps.push_back(afOp);
+    // Construct new indices as a remap of the old ones if a remapping has been
+    // provided. The indices of a memref come right after it, i.e.,
+    // at position memRefOperandPos + 1.
```
Reflow (use whole width).

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -92,119 +93,134 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
-    newMapOperands.push_back(extraIndex);
-  }
+    // Prepend 'extraIndices' in 'newMapOperands'.
+    for (auto extraIndex : extraIndices) {
```
Avoid using `auto` here - please spell out the type for readability.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -443,6 +460,16 @@ LogicalResult mlir::normalizeMemRef(AllocOp allocOp, Value *normalizedMemRef) {
   auto oldMemRef = allocOp.getResult();
   SmallVector<Value, 4> symbolOperands(allocOp.getSymbolicOperands());
 
+  // TODO(avarmapml) : Handle the initial filter to ensure that the only uses
+  //                   of memref is DeallocOp, CallOp, (maybe) ReturnOp
+  // if(llvm::any_of(oldMemRef.getUsers(),
+  //                 [](Operation *op) { 
+  //                   if(isMemRefDereferencingOp(op)) {
+  //                     return !(isa<DeallocOp>(op) || isa<CallOp>(op))
+  //                   }
+  //                 })) {
+  //   llvm::dbgs()<<"No going to support such"
+  // }
```
You can use `#if 0` /  `#endif` instead of commenting each line. In any case, please drop commented out code - you can keep the TODO, but store the commented out code separately for your reference if needed and not in this PR.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+
+      // Check whether argument is of MemRef type
+      // TODO(avarmapml): What about unranked memrefs? 
+      if(funcOp.getArgument(argIndex).getType().getKind() == StandardTypes::MemRef) {
```
Missing space after `if`. Please use clang-format.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+          continue;
+
+        auto layoutMaps = memrefType.getAffineMaps();
+        if (layoutMaps.size() != 1) {
```
Did you mean `layoutMaps.empty()`?

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        AffineMap layoutMap = layoutMaps.front();
+        
```
Drop extra new line.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+      results.push_back(ft.getResult(retIndex));
+    }
+
+    // Walk over each argument of a function to perform memref normalization (if any)
+    for (auto argIndex : llvm::seq<unsigned>(0, funcOp.getNumArguments())) {
```
Wrap to 80 columns. Use clang-format.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+}
+
+void ModuleNormalizeMemRef::runOnOperation() {
+  auto func = getOperation();
```
This is misnamed. `auto *module`?

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/ModuleNormalizeMemRef.cpp:
```diff
@@ -0,0 +1,214 @@
+          assert(ubConst.hasValue() && "should always have an upper bound");
+          if (ubConst.getValue() < 0)
+            // This is due to an invalid map that maps to a negative space.
+            // return failure();
```
Remove commented out code like these.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -269,10 +285,9 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
-      // no replacement in these cases.
-      return failure();
+    if (!isMemRefDereferencingOp(*op)) {
+      // As of now allowing any non-dereferencing type
```
Rephrase/fix comment please.

bondhugula reviewed (changes requested):
This is a fantastic start - thanks! A first round of superfluous comments that you may want to address.

bondhugula commented:
> Created a pass 'module-normalize-memref' to normalize memref layout for interprocedural calls.

The pass can just be named `-normalize-memrefs`

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -92,119 +93,134 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
+        state.attributes.push_back({namedAttr.first, newMapAttr});
+      } else {
+        state.attributes.push_back(namedAttr);
+      }
```
You can drop trivial braces here.

avarmapml committed:
[MLIR] Code cleanup for interprocedural memref layout normalization

-- Fixes for code review: indentation, variable naming, etc

avarmapml commented:
I've made the changes.
Also, since I'm using clang-format-6.0, accordingly the .clang-format file had to be changed as well.

bondhugula commented on mlir/.clang-format:
```diff
@@ -1,2 +1,2 @@
 BasedOnStyle: LLVM
-AlwaysBreakTemplateDeclarations: Yes
\ No newline at end of file
+AlwaysBreakTemplateDeclarations: true
```
You can drop this change from this PR.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to normalize interprocedural memref layouts.
```
-> "... implements an interprocedural pass to normalize memref layouts."

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+//===- NormalizeMemRefs.cpp ---------------------------------------===//
```
This pass can actually live in `lib/Transforms/` instead of in `lib/Dialects/Affine/Transforms` because it isn't specific to the affine dialect.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+/// ones with trivial identity layout ones.
+
+// Input :-
+// #pad = affine_map<(i) -> (i floordiv 4, i mod 4)>
```
Nit: naming - you can call this map `#tile` because it's not really padding.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+void NormalizeMemRefs::runOnOperation() {
+  auto module = getOperation();
```
Nit: `auto *module`.

avarmapml replied:
getOperation here doesn't have a return type 'Operation*'.
It instead returns 'mlir::ModuleOp' (after internally casting).
I've thus changed this line to :-
ModuleOp moduleOp = getOperation();

bondhugula replied:
I see. That's another reason we should spell out the type! :)

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+    OpBuilder b(funcOp);
+
+    FunctionType ft = funcOp.getType();
+    SmallVector<Type, 8> inputs;  // Holds the arguments of function.
```
Nit: avoid trailing comments - add on separate line.

bondhugula replied:
Please name this better `argMemRefTypes` or `inputMemRefTypes`?

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    FunctionType ft = funcOp.getType();
+    SmallVector<Type, 8> inputs;  // Holds the arguments of function.
+    SmallVector<Type, 4> results; // Holds the results of function.
```
Likewise.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    // Populating results with function's initial result (type) as
+    // this will be used to modify and set function's signature later.
+    for (auto retIndex : llvm::seq<unsigned>(0, funcOp.getNumResults())) {
```
Avoid trivial results. Avoid `auto` here for readability.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+    // Walk over each argument of a function to perform memref normalization (if
+    // any).
+    for (auto argIndex : llvm::seq<unsigned>(0, funcOp.getNumArguments())) {
+
```
Drop new line.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+      // TODO(avarmapml): What about unranked memrefs?
+      if (funcOp.getArgument(argIndex).getType().getKind() ==
+          StandardTypes::MemRef) {
+
```
Drop extra line.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+    for (auto argIndex : llvm::seq<unsigned>(0, funcOp.getNumArguments())) {
+
+      // Check whether argument is of MemRef type.
+      // TODO(avarmapml): What about unranked memrefs?
```
Unranked memrefs should be skipped. Since they rank itself isn't known, they will never be expected to have an affine map.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+
+      // Check whether argument is of MemRef type.
+      // TODO(avarmapml): What about unranked memrefs?
+      if (funcOp.getArgument(argIndex).getType().getKind() ==
+          StandardTypes::MemRef) {
```
You don't need to use `getKind()`, instead ```.getType().isa<MemRefType>()```

bondhugula replied:
Use early exit to avoid extra indentation for all the following code and thus making it simpler / easier to read, i.e., 
```
if (!funcOp.getArgument(argIndex).getType().isa<MemRefType>())
  continue;
// The rest of the code here will have one indentation less.
```

bondhugula replied:
Using `isa<MemRefType>()` will automatically take care filtering out unranked memrefs.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+          StandardTypes::MemRef) {
+
+        MemRefType memrefType =
+            funcOp.getArgument(argIndex).getType().cast<MemRefType>();
```
You can actually use a `dyn_cast` above (where you are checking whether it's a memref type). That way you'll get `memrefType`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+        if (rank == 0)
+          continue;
+
+        auto layoutMaps = memrefType.getAffineMaps();
```
Avoid `auto` and spell the type out for readability.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    // Walk over each argument of a function to perform memref normalization (if
+    // any).
+    for (auto argIndex : llvm::seq<unsigned>(0, funcOp.getNumArguments())) {
```
`auto` -> `unsigned`

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+          StandardTypes::MemRef) {
+
+        MemRefType memrefType =
+            funcOp.getArgument(argIndex).getType().cast<MemRefType>();
```
There are several places where you've used `funcOp.getArgument(argIndex).getType()` - assign it to `memrefArg` or `memrefArgType`?

avarmapml replied:
Since the type is not yet known, I'm assigning the same to `argType` instead.
Alsom I'm not using dyn_cast yet. Will handle that later.

bondhugula replied:
> Since the type is not yet known, I'm assigning the same to `argType` instead.
> Alsom I'm not using dyn_cast yet. Will handle that later.

An `isa` + a `cast` later suggests that a `dyn_cast` could be used. So, you could do:

```
MemRefType argMemRefType = ..getType().dyn_cast<MemRefType>();
if (!argMemRefType)
  continue;
...
// And now you have access to the MemRefType as well without needing a cast.
...
```

bondhugula replied:
https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates

for reference.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+        // Insert a new temporary argument with the new memref type.
+        funcOp.front().insertArgument(argIndex,
+                                      static_cast<Type>(newMemRefType));
+        auto newMemRef = funcOp.getArgument(argIndex);
```
Avoid `auto`. I understand some of the earlier code used `auto` a lot.

bondhugula replied:
Avoid `auto` - spell out type.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+                                            /*indexRemap=*/layoutMap,
+                                            /*extraOperands=*/{},
+                                            /*symbolOperands=*/{}))) {
+          // If it failed (due to escapes for example), bail out.
+          // Removing the temporary argument inserted previously.
```
Reflow to use width.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+        // Since in this pass the objective is to normalize arguments of a
+        // function and replace the uses accordingly, we check if the function
+        // return type uses the same old memref type.
+        // TODO(avarmapml): Check - A function's return type might have a
+        // different memref layout and a map.
```
It's not clear what this means - did you mean a non-identity layout map?

avarmapml replied:
So, the logic I've implemented checks if the return type has a memory layout that matches the old non-trivial memref layout in argument.
If there is a match an update (trivializing) is made to the return type.

But the return type might use a different memref layout (resulting from, say, alloc or a result of another function call) that should be normalized. I need to test these cases and accordingly handle them.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+      }
+    }
+
+    ArrayRef<Type> inp = makeArrayRef(inputs);
+    ArrayRef<Type> res = makeArrayRef(results);
```
`inp` <- name this better please.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -91,118 +90,131 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
-      affineApplyOps.push_back(afOp);
+  // The following checks if op is dereferencing memref and performs the access
+  // index rewrites.
+  if (isMemRefNonDereferencing == false) {
```
`if (!isMemRefNonDeferencing) {` to match style everywhere else.

bondhugula reviewed (changes requested):
Thanks for revising - posting more review comments 

Reg. getting changes from master, always use `rebase` instead of `merge`, i.e., `git rebase <origin/master>` for example.

bondhugula reviewed (changes requested):
Please also add test cases. I normally always start the review from the test cases!

bondhugula commented:
> Created a pass 'module-normalize-memref' to normalize memref layout for interprocedural calls.

You may want to rewrite the summary a bit. "interprocedural calls" isn't meaningful. Instead, "across function calls". Also, this can just be named `-normalize-memrefs`.

yashjain1129 commented on mlir/lib/Dialect/Affine/Transforms/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@ add_mlir_dialect_library(MLIRAffineTransforms
   SuperVectorize.cpp
   SimplifyAffineStructures.cpp
-  ModuleNormalizeMemRef.cpp
+  NormalizeMemRefs.cpp
```
Add NormalizeMemrefs.cpp alphabetically after loopUnrollAndJam.cpp

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+
+        // Insert a new temporary argument with the new memref type.
+        funcOp.front().insertArgument(argIndex,
+                                      static_cast<Type>(newMemRefType));
```
You don't need `static_cast<Type>`.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+                .setShape(newShape)
+                .setAffineMaps(b.getMultiDimIdentityMap(newRank));
+
+        auto oldMemRef = funcOp.getArgument(argIndex);
```
Avoid `auto` - spell out type.

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,220 @@
+          continue;
+        }
+
+        // Since in this pass the objective is to normalize arguments of a
```
-> ... normalize the layout maps of the memref arguments ...

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -69,10 +69,9 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
-    // Failure: memref used in a non-dereferencing context (potentially
-    // escapes); no replacement in these cases.
-    return failure();
+    isMemRefNonDereferencing = true;
```
To make it more intuitive, can you instead use `isMemRefDeferencing`?

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -93,118 +92,132 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
+      }
+    } else {
+      // No remapping specified.
+      remapOutputs.append(remapOperands.begin(), remapOperands.end());
```
Nit: `append` -> `assign`. (The former gives the impression that something is already in there, which isn't the case here.) I understand this already existed before.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -93,118 +92,132 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
+    newMap = simplifyAffineMap(newMap);
+    canonicalizeMapAndOperands(&newMap, &newMapOperands);
+    // Remove any affine.apply's that became dead as a result of composition.
+    for (auto value : affineApplyOps)
```
Avoid `auto` - spell out type.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -93,118 +92,132 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
+    // Remove any affine.apply's that became dead as a result of composition.
+    for (auto value : affineApplyOps)
+      if (value.use_empty())
+        value.getDefiningOp()->erase();
```
What if `getDefiningOp` is null, i.e., `value` is a block argument?

bondhugula replied:
Sorry, this is guaranteed from the logic above. We should have been asserting - but that's an issue with the pre-existing code.

bondhugula reviewed (changes requested):
For RAMUW, you can follow the structure:
```
if (!memrefDeferencing) {
  ..
  return
}

// Handle memref derefencing op.
```
That way all the white space change due to the extra indentation is completely gone and it's more readable with less indent.

avarmapml committed:
[MLIR] Shifted normalize-memrefs pass to lib/Transforms

-- The interprocedural pass to normalize memref layout maps is now shifted to lib/Transforms
-- Minor code fixes
-- A test case file for the pass has been added. More test cases to be added later.

avarmapml committed:
[MLIR] Added flag control to allow non-dereferencing ops in replaceAllMemRefUsesWith

-- This commit adds a flag control (default value 'false') that the caller of replaceAllMemRefUsesWith
   can use to specify whether non-dereferencing ops need to be replaced or not.

bondhugula commented on mlir/lib/Transforms/Utils/Utils.cpp:
```diff
@@ -69,11 +70,6 @@ LogicalResult mlir::replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
          newMemRef.getType().cast<MemRefType>().getElementType());
 
-  if (!isMemRefDereferencingOp(*op))
-    // Failure: memref used in a non-dereferencing context (potentially
-    // escapes); no replacement in these cases.
```
Please preserve these comments below.

bondhugula commented on mlir/test/Transforms/normalize-memrefs.mlir:
```diff
@@ -0,0 +1,22 @@
+// RUN: mlir-opt -allow-unregistered-dialect -normalize-memrefs %s | FileCheck %s
```
You don't need `-allow-unregistered-dialect`.

bondhugula commented on mlir/test/Transforms/normalize-memrefs.mlir:
```diff
@@ -0,0 +1,22 @@
+  return %A: memref<16xf64, #tile>
+}
+
+// CHECK-NEXT: affine.for %[[i:arg[0-9]+]]
+// CHECK-NEXT: %[[a:[0-9]+]] = affine.load %[[A]][%[[i]] floordiv 4, %[[i]] mod 4] : memref<4x4xf64>
+// CHECK-NEXT: %[[p:[0-9]+]] = mulf %[[a]], %[[a]] : f64
+// CHECK-NEXT: affine.store %[[p]], %[[A]][%[[i]] floordiv 4, %[[i]] mod 4] : memref<4x4xf64>
+// CHECK: return %[[A]] : memref<4x4xf64>
```
Please indent these for better readability (the part after the CHECK-*:).

bondhugula replied:
Missing line break at end of file.

bondhugula commented on mlir/test/Transforms/normalize-memrefs.mlir:
```diff
@@ -0,0 +1,22 @@
+// RUN: mlir-opt -allow-unregistered-dialect -normalize-memrefs %s | FileCheck %s
+
```
Add a comment here (a line or two) on what this file is testing.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+      unsigned newRank = layoutMap.getNumResults();
+      if (failed(fac.composeMatchingMap(layoutMap)))
+        continue;
+      // TODO: semi-affine maps.
```
semi-affine -> Handle semi-affine

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+
+namespace {
+
+/// All interprocedural memrefs with non-trivial layout maps are converted to
```
This whole comment and example can be moved up outside the anonymous namespace.

bondhugula reviewed (changes requested):
This looks really good now! Some more comments.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+  moduleOp.walk([&](FuncOp funcOp) {
+    OpBuilder b(funcOp);
+
+    FunctionType ft = funcOp.getType();
```
Comment here on what these are.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+      MemRefType memrefType;
+      Type argType = funcOp.getArgument(argIndex).getType();
+      // Check whether argument is of MemRef type.
+      if (!(memrefType = argType.dyn_cast<MemRefType>())) {
```
You can move the assignment outside the conditional on the declaration line, and then just do `if (!memrefType)` - would be simpler and same number of lines.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+        argTypes.push_back(argType);
+        continue;
+      }
+      // MemRefType memrefType = argType.cast<MemRefType>();
```
Remove commented out code.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+        continue;
+
+      ArrayRef<AffineMap> layoutMaps = memrefType.getAffineMaps();
+      if (layoutMaps.empty()) {
+        // No maps associated with this memref, simply add this as the
+        // function argument.
+        argTypes.push_back(memrefType);
+        continue;
+      }
+
+      AffineMap layoutMap = layoutMaps.front();
+      if (layoutMap == b.getMultiDimIdentityMap(rank)) {
+        // Memref has identity map. Already trivial, simply add this as the
+        // function argument.
+        argTypes.push_back(memrefType);
```
You can merge these two with: 

```
if (layoutMaps.empty() || layoutMaps.front() == b.getMultiDimIdentityMap(rank)) {
```
Merge the comments though.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+              .setShape(newShape)
+              .setAffineMaps(b.getMultiDimIdentityMap(newRank));
+
+      BlockArgument oldMemRef = funcOp.getArgument(argIndex);
```
Move this below closer to its first use.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+      // TODO(avarmapml): Check - A function's return type might have a
+      // different memref layout and a map.
+      for (unsigned retIndex : llvm::seq<unsigned>(0, funcOp.getNumResults())) {
+        if (resultTypes[retIndex] == memrefType) {
+          resultTypes[retIndex] = newMemRef.getType();
+        }
```
Drop trivial braces.

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+      // All uses for the argument with old memref type were replaced
+      // successfully. So we remove the old argument now.
+      // TODO(avarmapml): replaceAllUsesWith.
+      funcOp.front().eraseArgument(argIndex + 1);
```
Shouldn't this be `funcOp.eraseArgument`?

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+    ArrayRef<Type> results = makeArrayRef(resultTypes);
+    // We create a new function type and modify the function signature with this
+    // new type.
+    FunctionType newFT = FunctionType::get(/*inputs=*/inputs,
```
Avoid FT (not camelBack) - instead something like `newFuncType`.

bondhugula reviewed (changes requested):
Some more comments.

bondhugula commented on mlir/test/Transforms/normalize-memrefs.mlir:
```diff
@@ -0,0 +1,22 @@
+#tile = affine_map<(i) -> (i floordiv 4, i mod 4)>
+func @matmul(%A: memref<16xf64, #tile>, %B: index, %C: memref<8xf64>) -> (memref<16xf64, #tile>) {
+  affine.for %arg3 = 0 to 16 {
+        %a = affine.load %A[%arg3] : memref<16xf64, #tile>
```
Use two space indent.

bondhugula commented on mlir/test/Transforms/normalize-memrefs.mlir:
```diff
@@ -0,0 +1,22 @@
+// RUN: mlir-opt -allow-unregistered-dialect -normalize-memrefs %s | FileCheck %s
+
+// CHECK-LABEL: func @matmul
```
You can move these below - to right above the function.

avarmapml replied:
I didn't understand this.
Am I to bring all the 'CHECK*'s right above the function?
OR
Just the 'CHECK-LABEL' part should be right above the function?
(I have used 'CHECK-SAME' that logically works with 'CHECK-LABEL', so I presume that needs to be moved as well)

bondhugula replied:
I meant to just move these two CHECK lines to right above the function, which would be

```
// CHECK-LABEL: func @matmul
// CHECK-SAME: (%[[A:arg[0-9]+]]: memref<4x4xf64>, %[[B:arg[0-9]+]]: index, %[[C:arg[0-9]+]]: memref<8xf64>) -> memref<4x4xf64>
func @matmul(%A: memref<16xf64, #tile>, %B: index, %C: memref<8xf64>) -> (memref<16xf64, #tile>) {
```

avarmapml replied:
Oh okay, cool, thank you!

bondhugula commented on mlir/lib/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -0,0 +1,221 @@
+void NormalizeMemRefs::runOnOperation() {
+  ModuleOp moduleOp = getOperation();
+
+  moduleOp.walk([&](FuncOp funcOp) {
```
I didn't notice this earlier: you could just outline this lambda. Please create a `runOnFunction(FuncOp funcOp)` and outline this out. A lambda function this long should really be a separate function for better readability. You also get rid of the extra indent/nesting that way.

avarmapml replied:
Have outlined the lambda function into runOnFunction(FuncOp funcOp).
How to refactor the code so that AllocOp and a function argument can use the same code needs a little thought. I'd look into that for sure because the readability has indeed improved after the said changes and I'm looking forward to bettering it.

bondhugula replied:
> Have outlined the lambda function into runOnFunction(FuncOp funcOp).
> How to refactor the code so that AllocOp and a function argument can use the same code needs a little thought. I'd look into that for sure because the readability has indeed improved after the said changes and I'm looking forward to bettering it.

Sounds great. You can start with moving that `normalizeMemRef` in this file (right next to runOnFunction) and we can see how to refactor it to share code. It should be straightforward if we take a crack.

avarmapml replied:
Sure, sir. Will look into this.
`HigherOrderPolyhedralOpt.cpp` uses `normalizeMemRef`, will have to change the the path and the namespace.

bondhugula replied:
> Sure, sir. Will look into this.
> `HigherOrderPolyhedralOpt.cpp` uses `normalizeMemRef`, will have to change the the path and the namespace.

I see - good catch. Note that HOPT doesn't exist upstream. We may still want to have that callable utility - we need to see how to refactor it so that it can be used from the pass as well as separately for alloc ops.

bondhugula reviewed (changes requested):
Abhishek, since we want to upstream this, the patch shouldn't introduce duplicate code. Can you thus move all of the existing normalize memref test cases from `simplify-affine-structures.mlir` into your test file and move the old normalizeMemRefs into this file? (i.e., make sure your pass subsumes it and remove the call to normalizeMemref from SimplifyAffineStructures).

bondhugula commented:
> Abhishek, since we want to upstream this, the patch shouldn't introduce duplicate code. Can you thus move all of the existing normalize memref test cases from `simplify-affine-structures.mlir` into your test file and move the old normalizeMemRefs into this file? (i.e., make sure your pass subsumes it and remove the call to normalizeMemref from SimplifyAffineStructures).

We can then refactor to make sure the rewriting for the `allocOp` part shares the same code as with the function argument case.

avarmapml committed:
[MLIR] Normalize non-trivial memref layout

-- This commit normalizes the AllocOps and interprocedural memref map layouts using a single pass
   'normalize-memrefs'
-- Involves a few minor code refactoring
-- All the test cases to check memref layout normalization is in a single file

avarmapml commented:
PR moved to another repo.
